rust/lance/src/io/exec/filtered_read.rs

```diff
@@ -0,0 +1,1584 @@
+    /// the row address.
+    pub fn with_projection(mut self, projection: Projection) -> Self {
+        self.projection = projection;
+        self
+    }
+}
+
+/// A plan node that reads a dataset, applying an optional filter and projection.
+///
+/// This node may execute a scan or it may execute a take.  By default, it picks the best
+/// approach based the expected query cost which is determined by:
+///  - Size of data in desired columns
+///  - Number of rows matching the index search
+///  - Filesystem parameters (e.g. block size)
+///
+/// This decision is made during execution, after the index search is complete, and not during
+/// planning.
+///
+/// In the future, we may introduce high-level cardinality statistics similar to those used by query
+/// engines like Postgres.  This might allow us to know, without executing an index search, that a scan
+/// would be better.  In that case we accept the force_scan hint to skip the index search.
+#[derive(Debug)]
+pub struct FilteredReadExec {
+    dataset: Arc<Dataset>,
+    options: FilteredReadOptions,
+    properties: PlanProperties,
+    metrics: ExecutionPlanMetricsSet,
+    // When execute is first called we will initialize the FilteredReadStream.  In order to support
+    // multiple partitions, each partition will share the stream.
+    running_stream: Arc<AsyncMutex<Option<FilteredReadStream>>>,
+}
+
+impl FilteredReadExec {
+    pub fn try_new(dataset: Arc<Dataset>, mut options: FilteredReadOptions) -> Result<Self> {
+        if options.with_deleted_rows {
+            // Ensure we have the row id column if with_deleted_rows is set
+            options.projection = options.projection.with_row_id();
+        }
+
+        if options.projection.is_empty() {
+            return Err(Error::InvalidInput {
+                source:
+                    "no columns were selected and with_row_id / with_row_address is false, there is nothing to scan"
+                        .into(),
+                location: location!(),
+            });
+        }
+
+        if options.scan_range_after_filter.is_some() {
+            // TODO: Add support for this.  It isn't too tricky to apply the limit when we have a scalar index.
+            // However, we need to make sure and apply the limit with unindexed fragments too.  This is also not
+            // too bad...unless there are multiple partitions.  In that case we have several readers reading in
+            // parallel and applying the limit after the fact is tricky.
+            return Err(Error::NotSupported {
+                source: "scan_range_after_filter not yet implemented"
+                    .to_string()
+                    .into(),
+                location: location!(),
+            });
+        }
+        let output_schema = Arc::new(options.projection.to_arrow_schema()?);
+
+        let properties = PlanProperties::new(
+            EquivalenceProperties::new(output_schema),
+            Partitioning::RoundRobinBatch(1),
```

Comment:
Are multiple partitions hooked up at all yet?

Reply:
Nope.  This has evolved slightly in a future PR but I don't want to dive into multiple partitions just yet.  If we go that route we need to have multiple partitions throughout the exec plan and may need things like a special accumulation node that reads from multiple partitions and builds up a unified roaring bitmap in parallel (e.g. to feed input as a prefilter)